File: 200/最佳植树距离/最佳植树距离.py
# https://blog.csdn.net/qfc_128220/article/details/130633638
from typing import *

class Solution:
    def maxDistance(self, position: List[int], m: int) -> int:
        position.sort()
        min_dis = 1  # 最小植树间隔就是相邻位置
        max_dis = position[-1] - position[0] # 最大植树间隔就是两端位置
        res = -1

        #检查选定的最小植树距离是否合法
        def check(x: int) -> bool:
            cnt = 1
            pre = position[0]
            
            # 间隔x放一个树，看一共能放几个
            for i in range(1, len(position)):
                if position[i] - pre >= x:
                    pre = position[i]
                    cnt += 1
            return cnt >= m

        # 二分查找从最小值到最大值中最大的能满足的
        while min_dis <= max_dis:
            mid = (max_dis - min_dis) // 2 + min_dis
            if check(mid):
                res = mid
                min_dis = mid + 1
            else:
                max_dis = mid - 1

        return res
            

File: 200/最佳植树距离/最佳植树距离.cpp
#include <iostream>
#include <string>
#include <vector>

std::vector<int> split(const std::string& str, char delimiter) {
    std::vector<int> result;
    size_t pos = 0;
    size_t found = 0;
    while ((found = str.find(delimiter, pos)) != std::string::npos) {
        result.push_back(std::stoi(str.substr(pos, found - pos)));
        pos = found + 1;
    }
    result.push_back(std::stoi(str.substr(pos)));
    return result;
}

bool check(int dis, const std::vector<int>& pos, int trees) {
    if (pos.empty()) {
        return false;
    }
    int count = 0;
    int pre = pos[0];
    for (int i = 1; i < pos.size(); i++) {
        if (pos[i] - pre >= dis) {
            count++;
            pre = pos[i];
        }
    }

    return count >= trees;
}

int get_res(const std::vector<int>& pos, int trees) {
    if (pos.size() < 2) {
        return 0;
    }
    int min_dis = 1;
    int max_dis = pos[pos.size()  - 1] - pos[0];
    int res = 0;
    while (min_dis <= max_dis) {
        int mid = ((max_dis - min_dis) >> 1) + min_dis;
        if (check(mid, pos, trees)) {
            res = mid;
            min_dis = mid + 1;
        } else {
            max_dis = mid - 1;
        }
    }
    return res;
}

int main() {
    int n;
    scanf("%d\n", &n);
    std::string line;
    getline(std::cin, line);
    std::vector<int> pos = split(line, ' ');
    int trees;
    std::cin >> trees;
    std::cout << get_res(pos, trees) << std::endl;
}

/*
7
1 5 3 6 10 7 13
3
*/